From 5bbc8047f8e0211e9503cf538b16a192c5f98866 Mon Sep 17 00:00:00 2001
From: Cheng-Cheng Lo <locc@google.com>
Date: Thu, 21 Aug 2025 16:50:01 +0800
Subject: [PATCH] Disallow factory reset while in DSU mode

Bug: 430568718
Flag: EXEMPT CVE_FIX
Test: presubmit
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:72f795b3bcb0c98971f5f15689c9399c8160bfcb)
Merged-In: I36afa074f16bb7ebf8b617de73dab3d4df5fb5dc
Change-Id: I36afa074f16bb7ebf8b617de73dab3d4df5fb5dc
---
 core/java/android/os/RecoverySystem.java | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/core/java/android/os/RecoverySystem.java b/core/java/android/os/RecoverySystem.java
index a3b836adfc8b..659816497bfe 100644
--- a/core/java/android/os/RecoverySystem.java
+++ b/core/java/android/os/RecoverySystem.java
@@ -38,6 +38,7 @@ import android.content.IntentFilter;
 import android.content.IntentSender;
 import android.content.pm.PackageManager;
 import android.hardware.display.DisplayManager;
+import android.os.image.DynamicSystemManager;
 import android.provider.Settings;
 import android.telephony.SubscriptionInfo;
 import android.telephony.SubscriptionManager;
@@ -958,9 +959,15 @@ public class RecoverySystem {
     public static void rebootWipeUserData(Context context, boolean shutdown, String reason,
             boolean force, boolean wipeEuicc) throws IOException {
         UserManager um = (UserManager) context.getSystemService(Context.USER_SERVICE);
+        DynamicSystemManager dsuManager = (DynamicSystemManager)
+                    context.getSystemService(Context.DYNAMIC_SYSTEM_SERVICE);
         if (!force && um.hasUserRestriction(UserManager.DISALLOW_FACTORY_RESET)) {
             throw new SecurityException("Wiping data is not allowed for this user.");
         }
+
+        if (dsuManager.isInUse()) {
+            throw new SecurityException("Wiping data is not allowed while in DSU mode.");
+        }
         final ConditionVariable condition = new ConditionVariable();
 
         Intent intent = new Intent("android.intent.action.MASTER_CLEAR_NOTIFICATION");
-- 
2.52.0.rc1.455.g30608eb744-goog

